presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPartitionInsertingPageSource.java

```diff
@@ -0,0 +1,204 @@
+ * are not partition or metadata fields.
+ * <p>
+ * When a new page is requested, this class inserts the partition fields as
+ * RLE-encoded blocks to save memory and IO cost at runtime.
+ */
+public class IcebergPartitionInsertingPageSource
+        implements ConnectorPageSource
+{
+    private final IcebergColumnHandle[] nonPartitionColumnIndexes;
+    private final ConnectorPageSourceWithRowPositions delegateWithPositions;
+    private final ConnectorPageSource delegate;
+    private final Block[] partitionValueBlocks;
+    private final int[] outputIndexes;
+
+    public IcebergPartitionInsertingPageSource(
+            List<IcebergColumnHandle> fullColumnList,
+            Map<Integer, Object> metadataValues,
+            Map<Integer, HivePartitionKey> partitionKeys,
+            Function<List<IcebergColumnHandle>, ConnectorPageSourceWithRowPositions> delegateSupplier)
+    {
+        this.nonPartitionColumnIndexes = new IcebergColumnHandle[fullColumnList.size()];
+
+        int delegateIndex = 0;
+        // maps output array index to index of input page from delegate provider.
+        this.outputIndexes = new int[fullColumnList.size()];
+        // generate array of non-partition column indexes
+        for (int i = 0; i < fullColumnList.size(); i++) {
+            IcebergColumnHandle handle = fullColumnList.get(i);
+            if (handle.getColumnType() == PARTITION_KEY ||
+                    partitionKeys.containsKey(handle.getId()) ||
+                    isMetadataColumnId(handle.getId())) {
+                // is partition key, don't include for delegate supplier
+                continue;
+            }
+            nonPartitionColumnIndexes[i] = handle;
+            outputIndexes[i] = delegateIndex;
+            delegateIndex++;
+        }
+        List<IcebergColumnHandle> delegateColumns = Arrays.stream(nonPartitionColumnIndexes)
+                .filter(Objects::nonNull)
+                .collect(toImmutableList());
+
+        this.partitionValueBlocks = IntStream.range(0, fullColumnList.size())
+                .mapToObj(idx -> {
+                    IcebergColumnHandle column = fullColumnList.get(idx);
+                    if (nonPartitionColumnIndexes[idx] != null) {
+                        return null;
+                    }
+
+                    if (partitionKeys.containsKey(column.getId())) {
+                        HivePartitionKey icebergPartition = partitionKeys.get(column.getId());
+                        Type type = column.getType();
+                        Object prefilledValue = deserializePartitionValue(type, icebergPartition.getValue().orElse(null), column.getName());
+                        return nativeValueToBlock(type, prefilledValue);
+                    }
+                    else if (column.getColumnType() == PARTITION_KEY) {
+                        // Partition key with no value. This can happen after partition evolution
+                        Type type = column.getType();
```

Comment:
nit: we can move this before the if..else and reuse or directly use `column.getType()`.